ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -973,11 +1090,28 @@ private void handleMetadataBlob(BlobData blobData) throws IOException, MessageFo
           numChunksTotal = 0;
           dataChunks = null;
         } else {
-          chunkIdIterator = keys.listIterator();
-          numChunksTotal = keys.size();
-          dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
-          for (int i = 0; i < dataChunks.length; i++) {
-            dataChunks[i] = new GetChunk(chunkIdIterator.nextIndex(), (BlobId) chunkIdIterator.next());
+          // if blob is encrypted, then decryption is required only in case of GetBlobInfo and GetBlobAll (since user-metadata
+          // is expected to be encrypted). Incase of GetBlob, Metadata blob does not need any decryption even if BlobProperties says so
+          if (getOperationFlag() != MessageFormatFlags.Blob && blobInfo.getBlobProperties().isEncrypted()) {
+            decryptionRequired.set(true);
+            logger.trace("Submitting decrypt job for Metadaata chunk {}", blobId);
+            long startTimeMs = System.currentTimeMillis();
+            cryptoJobHandler.submitJob(
+                new DecryptJob(blobId, encryptionKey, null, ByteBuffer.wrap(blobInfo.getUserMetadata()), cryptoService,
+                    kms, (DecryptJob.DecryptJobResult result, Exception exception) -> {
+                  routerMetrics.contentDecryptTimeMs.update(System.currentTimeMillis() - startTimeMs);
+                  logger.trace("Handling decrypt job call back for Metadata chunk {} to set decrypt callback results",
+                      blobId);
+                  decryptCallbackResultInfo.setResultAndException(result, exception);
+                  routerCallback.onPollReady();
+                }));
+          } else {
+            chunkIdIterator = keys.listIterator();
+            numChunksTotal = keys.size();
+            dataChunks = new GetChunk[Math.min(keys.size(), NonBlockingRouter.MAX_IN_MEM_CHUNKS)];
```

Comment:
please make sure this method is not duplicating things (or the other method).

Reply:
yes. as I told you, poll() in line 248 is making use of this. So depending on whether its encrypted or not, had to do it this way